src/core/btree_set.h

```diff
@@ -0,0 +1,281 @@
+    -> std::pair<BTreeNode*, Key_t> {
+  assert(path.Depth() > 0u);
+
+  BTreeNode* node = path.Last().first;
+  assert(node->IsLeaf() && node->AvailableSlotCount() == 0);
+
+  unsigned insert_pos = path.Last().second;
+  unsigned level = path.Depth() - 1;
+  if (level > 0) {
+    BTreeNode* parent = path.Node(level - 1);
+    unsigned pos = path.Position(level - 1);
+    assert(parent->Child(pos) == node);
+
+    std::pair<BTreeNode*, unsigned> rebalance_res = parent->RebalanceChild(pos, insert_pos);
+    if (rebalance_res.first) {
+      rebalance_res.first->LeafInsert(rebalance_res.second, item);
+      return {nullptr, 0};
+    }
+  }
+
+  Key_t median;
+  BTreeNode* right = CreateNode(node->IsLeaf());
+  node->Split(right, &median);
+
+  assert(node->NumItems() < BTreeNode::kMaxLeafKeys);
+
+  if (insert_pos <= node->NumItems()) {
+    assert(item < median);
+    node->LeafInsert(insert_pos, item);
+  } else {
+    assert(item > median);
+    right->LeafInsert(insert_pos - node->NumItems() - 1, item);
+  }
+
+  // we now must add right to the paren if it exists.
+  while (level-- > 0) {
+    node = path.Node(level);            // level up, now node is parent.
+    insert_pos = path.Position(level);  // insert_pos is position of node in parent.
+
+    assert(!node->IsLeaf() && insert_pos <= node->NumItems());
+
+    if (node->NumItems() == BTreeNode::kMaxInnerKeys) {
+      if (level > 0) {
+        BTreeNode* parent = path.Node(level - 1);
+        unsigned node_pos = path.Position(level - 1);
+        assert(parent->Child(node_pos) == node);
+        std::pair<BTreeNode*, unsigned> rebalance_res =
+            parent->RebalanceChild(node_pos, insert_pos);
+        if (rebalance_res.first) {
+          rebalance_res.first->InnerInsert(rebalance_res.second, median, right);
+          return {nullptr, 0};
+        }
+      }
+
+      Key_t parent_median;
+      BTreeNode* parent_right = CreateNode(false);
+      node->Split(parent_right, &parent_median);
+      assert(node->NumItems() < BTreeNode::kMaxInnerKeys);
+
+      if (insert_pos <= node->NumItems()) {
+        assert(median < parent_median);
+        node->InnerInsert(insert_pos, median, right);
+      } else {
+        assert(median > parent_median);
+        parent_right->InnerInsert(insert_pos - node->NumItems() - 1, median, right);
+      }
+      right = parent_right;
+      median = parent_median;
+    } else {
+      node->InnerInsert(insert_pos, median, right);
+      return {nullptr, 0};
+    }
+  }
+
+  return {right, median};
+}
+
+template <typename T, typename Policy>
+detail::BTreeNode<Policy>* BTree<T, Policy>::CreateNode(bool leaf) {
+  num_nodes_++;
+  void* ptr = mr_->allocate(BTreeNode::kTargetNodeSize, 8);
+  BTreeNode* node = new (ptr) BTreeNode(leaf);
+
+  return node;
+}
+
+template <typename T, typename Policy> void BTree<T, Policy>::DestroyNode(BTreeNode* node) {
+  void* ptr = node;
+  mr_->deallocate(ptr, BTreeNode::kTargetNodeSize, 8);
+  num_nodes_--;
+}
```

Comment:
I'd leave a high level comment somewhere above BTreeNode to explain why it offsets its this pointer beyond its size. Shahar is right, we could just keep the buffer in the struct to occupy this space, but it would make implementing point 3 on your todo (smaller nodes for small trees) more difficult and you'd have to revert back to this approach

Reply:
Exactly. absl implements small object optimization. In short, the answer is - I embrace the fact that high-level languages can not support dynamically created structs. We usually do not use this approach but when it's needed I am not going to avoid it because it's "not nice". And btw, I used a similar approach in dense_set when allocating metadata for expire time of hash elements.